### PR TITLE
S3 - Register Form

### PR DESCRIPTION
**Pull request by Ashton-Jeffrey**
- Created register form
- Validation for password, username, email. 
- Once registered it redirects to index page
- Hiding of register in nav once logged in.

### CODE REVIEW

File: .idea/php.xml
Comment (mporam) [inline .idea/php.xml:1]: the .idea folder needs to be in your .gitignore

File: index.php [hunk 1]
```diff
@@ -7,6 +7,13 @@
 $db = connectToDb();
 $blogModel = new BlogModel($db);
 $blogs = $blogModel->getAllPosts();
+
+if (isset($_SESSION['userid'])){
+    $isUserLoggedIn = '';
+}
+else {
```
Comment (mporam) [inline index.php]: PSR-12 says these should be on the same line:
`} else {`

File: index.php [hunk 2]
```diff
@@ -22,7 +29,8 @@
     <div class="flex gap-5">
         <?php echo isset($_SESSION['userid']) ?
         '<a href="logout.php">Logout</a>':
-        '<a href="login.php">Login</a>'; ?>
+        '<a href="login.php">Login</a>
+        <a href="register.php">Register</a>';?>
```
Comment (mporam) [inline index.php:27]: this if statement is doing the same as the one above? this is the better way to do it, remove the one above

File: index.php [hunk 3]
```diff
@@ -22,7 +29,8 @@
...
     </div>
 </nav>
 <!--<form class="container lg:w-1/2 gap-5 mx-auto mb-10 flex justify-between items-center flex-col lg:flex-row px-5 sm:px-0">-->
```
Comment (mporam) [inline index.php]: dont leave commented code in, please remove this

File: index.php [hunk 4]
```diff
@@ -51,6 +59,7 @@
 <!--    </div>-->
 <!--    <input class="px-3 py-2 text-lg bg-indigo-400 hover:bg-indigo-700 hover:text-white transition inline-block rounded-sm" type="submit" value="Filter">-->
 <!--</form>-->
```
Comment (mporam) [inline index.php]: again, remove this

File: register.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+$db = connectToDb();
+session_start();
+$userError = '';
+$passwordError = '';
+$emailError = '';
+
+$userModel = new UserModel($db);
+
+// Check if the form is submitted
+if ($_SERVER["REQUEST_METHOD"] == "POST") {
```
Comment (mporam) [inline register.php]: this only checks that your making a post request, not that the keys `username`, `email` or `password` actually exist. lines 17-19 will error if they dont, so its better to do a `!empty()` on all 3 at this point (then you dont need to check its a post request)

File: register.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    // Get form data
+    $username = $_POST["username"];
+    $email = $_POST["email"];
+    $password = $_POST["password"];
+
+    if ($userModel->usernameExists($username)) {
+        $userError = 'This user already exists';
+    };
```
Comment (mporam) [inline register.php]: this is a security vulnerability. this allows malicious people to discover the usernames of people in your system

File: register.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    if (isset($_POST["password"])) {
+        $password = $_POST["password"];
```
Comment (mporam) [inline register.php]: youv already done this on line 19

File: register.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    // Get form data
+    $username = $_POST["username"];
+    $email = $_POST["email"];
+    $password = $_POST["password"];
+
+    if ($userModel->usernameExists($username)) {
+        $userError = 'This user already exists';
+    };
+
+    if (isset($_POST["password"])) {
```
Comment (mporam) [inline register.php]: what happens if there is no password?

File: register.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        $password = $_POST["password"];
+        $passwordError = passwordHint($password);
+    }
+    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+        $emailError = 'Invalid email format';
+    }
+
+    // Call registerUser function
+    $result = $userModel->registerUser($username, $email, $password);
```
Comment (mporam) [inline register.php]: if your validation above fails, it is still going to run this function and register the user??

File: register.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    $user = $userModel->getUserByEmail($email);
+    $_SESSION['userid'] = $user->id;
+    header('Location: index.php');
+    // Display result
+    echo "<p>$result</p>";
```
Comment (mporam) [inline register.php]: this is pointless, youv redirected on the line above

File: register.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        $userError = 'This user already exists';
+    };
+
+    if (isset($_POST["password"])) {
+        $password = $_POST["password"];
+        $passwordError = passwordHint($password);
+    }
+    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+        $emailError = 'Invalid email format';
+    }
```
Comment (mporam) [inline register.php]: you have an email class for exactly this purpose, why duplicate effort?

File: src/Models/UserModel.php [hunk 12]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
     {
         $this->db = $db;
     }
-    public function getUserByEmail(string $email): User | false
+
+    public function registerUser($username, $email, $password)
```
Comment (mporam) [inline src/Models/UserModel.php]: missing type hints

File: src/Models/UserModel.php [hunk 13]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
...
+    {
+        $pdo = $this->db;
+        //hash the password that was provided
+        $hashedPassword = password_hash($password, PASSWORD_BCRYPT);
```
Comment (mporam) [inline src/Models/UserModel.php]: dont use BCRYPT, use `PASSWORD_DEFAULT`

File: src/Models/UserModel.php [hunk 14]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
...
+
+        $query = $pdo->prepare('INSERT INTO `users` (`username`, `email`, `password`) VALUES(:username, :email, :hashedPassword)');
+
+        // Execute the query
+        $query->execute([
```
Comment (mporam) [inline src/Models/UserModel.php]: what happens if this query fails?

File: src/Models/UserModel.php [hunk 15]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
...
+            ':username' => $username,
+            ':email' => $email,
+            ':hashedPassword' => $hashedPassword,
+        ]);
+    }
+    public function usernameExists($username): bool
```
Comment (mporam) [inline src/Models/UserModel.php]: missing param typehint

File: src/Models/UserModel.php [hunk 16]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
     {
         $this->db = $db;
     }
-    public function getUserByEmail(string $email): User | false
+
+    public function registerUser($username, $email, $password)
+    {
+        $pdo = $this->db;
```
Comment (mporam) [inline src/Models/UserModel.php]: why? this is a waste of a line, just use `$this->pdo`

File: src/Models/UserModel.php [hunk 17]
```diff
@@ -9,7 +9,37 @@ public function __construct(PDO $db)
...
+        // Execute the query
+        $query->execute([
+            ':username' => $username,
+            ':email' => $email,
+            ':hashedPassword' => $hashedPassword,
+        ]);
+    }
+    public function usernameExists($username): bool
+    {
+        $pdo = $this->db;
```
Comment (mporam) [inline src/Models/UserModel.php]: again, this is unnecessary

File: tests/RegisterTest.php [hunk 18]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+require_once 'src/passwordHint.php';
+
+use PHPUnit\Framework\TestCase;
+
+class RegisterTest extends TestCase
+{
+    public function test_passwordHintError()
```
Comment (mporam) [inline tests/RegisterTest.php:9]: no test for what happens if the password is atleast 8 characters long

**Review by mporam — CHANGES_REQUESTED** [review]

**Comment by Ashton-Jeffrey** [comment]
Thanks Mike reviewed the above and pushed to S3. Ready to review again please.

File: register.php [hunk 19]
```diff
@@ -12,32 +12,29 @@
 $userModel = new UserModel($db);
 
 // Check if the form is submitted
-if ($_SERVER["REQUEST_METHOD"] == "POST") {
+if(!empty($_POST['submit'])) {
```
Comment (mporam) [inline register.php]: close, but you need to check the username, email and password like:
```
if (!empty($_POST['username']) && !empty($_POST['password']) && !empty($_POST['email'])) {
```

File: register.php [hunk 20]
```diff
@@ -12,32 +12,29 @@
...
 
+    $newEmail = new Email($email);
+
     if ($userModel->usernameExists($username)) {
-        $userError = 'This user already exists';
-    };
+        $userError = 'Sorry this is not a valid username';
+    } else { $userError = '';}
 
     if (isset($_POST["password"])) {
```
Comment (mporam) [inline register.php]: do this above as mentioned and remove if statement

File: src/Models/UserModel.php [hunk 21]
```diff
@@ -10,26 +10,29 @@ public function __construct(PDO $db)
         $this->db = $db;
     }
 
-    public function registerUser($username, $email, $password)
+    public function registerUser( string $username,  Email $email, string $password)
```
Comment (mporam) [inline src/Models/UserModel.php]: missing return type

**Review by mporam — CHANGES_REQUESTED** [review]

**Comment by Ashton-Jeffrey** [comment]
Thanks Mike reviewed and ready for review please

File: register.php [hunk 22]
```diff
@@ -24,9 +24,7 @@
         $userError = 'Sorry this is not a valid username';
     } else { $userError = '';}
 
-    if (isset($_POST["password"])) {
-        $passwordError = passwordHint($password);
-    }
+    $passwordError = passwordHint($password);
 
     if (empty($userError) && empty($passwordError) && empty($emailError)) {
```
Comment (mporam) [inline register.php:33]: $emailError will always be emty as you have nothing that can set it anymore. your check needs to lok at if you have a valid Email object inside $newEmail

**Review by mporam — CHANGES_REQUESTED** [review]

**Comment by Ashton-Jeffrey** [comment]
Thanks Mike updated and ready for review please

**Review by mporam — APPROVED** [review]

**Pull request merged**